File: mobile/test/data/models/daily_plan_test.dart
/// Tests for daily_plan.dart (Phase 6 mobile).
import 'package:flutter_test/flutter_test.dart';
import 'package:alfanumrik/data/models/daily_plan.dart';
import 'package:alfanumrik/data/models/goal_profile.dart';

void main() {
  group('DailyPlanItemKind.fromString', () {
    test('resolves all 6 kinds', () {
      expect(DailyPlanItemKind.fromString('pyq'), DailyPlanItemKind.pyq);
      expect(DailyPlanItemKind.fromString('concept'), DailyPlanItemKind.concept);
      expect(DailyPlanItemKind.fromString('practice'), DailyPlanItemKind.practice);
      expect(DailyPlanItemKind.fromString('challenge'), DailyPlanItemKind.challenge);
      expect(DailyPlanItemKind.fromString('review'), DailyPlanItemKind.review);
      expect(DailyPlanItemKind.fromString('reflection'), DailyPlanItemKind.reflection);
    });
    test('returns null for null/unknown', () {
      expect(DailyPlanItemKind.fromString(null), isNull);
      expect(DailyPlanItemKind.fromString('foo'), isNull);
    });
  });

  group('DailyPlanItem.fromJson', () {
    test('parses a complete item', () {
      final item = DailyPlanItem.fromJson({
        'kind': 'pyq',
        'titleEn': 'PYQ daily streak',
        'titleHi': 'PYQ पुराने प्रश्न',
        'estimatedMinutes': 20,
        'rationale': 'goal=board_topper',
      });
      expect(item.kind, DailyPlanItemKind.pyq);
      expect(item.titleEn, 'PYQ daily streak');
      expect(item.estimatedMinutes, 20);
    });

    test('throws on unknown kind', () {
      expect(
        () => DailyPlanItem.fromJson({'kind': 'foo'}),
        throwsArgumentError,
      );
    });
  });

  group('DailyPlan.fromJson', () {
    test('decodes a board_topper plan payload', () {
      final plan = DailyPlan.fromJson({
        'goal': 'board_topper',
        'totalMinutes': 45,
        'generatedAt': '2026-05-04T08:30:00.000Z',
        'items': [
          {'kind': 'pyq', 'titleEn': 'PYQ', 'titleHi': 'पुराने', 'estimatedMinutes': 20, 'rationale': 'x'},
          {'kind': 'practice', 'titleEn': 'HOTS', 'titleHi': 'HOTS', 'estimatedMinutes': 15, 'rationale': 'x'},
          {'kind': 'review', 'titleEn': 'Check', 'titleHi': 'जाँच', 'estimatedMinutes': 5, 'rationale': 'x'},
          {'kind': 'reflection', 'titleEn': 'Note', 'titleHi': 'नोट', 'estimatedMinutes': 5, 'rationale': 'x'},
        ],
      });
      expect(plan.goal, GoalCode.boardTopper);
      expect(plan.totalMinutes, 45);
      expect(plan.items.length, 4);
      expect(plan.items.first.kind, DailyPlanItemKind.pyq);
      expect(plan.isEmpty, false);
    });

    test('decodes empty plan', () {
      final plan = DailyPlan.fromJson({
        'goal': null,
        'totalMinutes': 0,
        'items': [],
        'generatedAt': '2026-05-04T08:30:00.000Z',
      });
      expect(plan.goal, isNull);
      expect(plan.isEmpty, true);
    });

    test('handles missing fields', () {
      final plan = DailyPlan.fromJson({});
      expect(plan.goal, isNull);
      expect(plan.totalMinutes, 0);
      expect(plan.items, isEmpty);
    });

    test('DailyPlan.empty factory', () {
      final p = DailyPlan.empty();
      expect(p.goal, isNull);
      expect(p.items, isEmpty);
    });
  });

  group('DailyPlanResponse.fromJson', () {
    test('decodes full envelope', () {
      final r = DailyPlanResponse.fromJson({
        'success': true,
        'flagEnabled': true,
        'data': {
          'goal': 'olympiad',
          'totalMinutes': 60,
          'items': [
            {'kind': 'challenge', 'titleEn': 'C', 'titleHi': 'चुनौती', 'estimatedMinutes': 30, 'rationale': 'x'},
          ],
          'generatedAt': '2026-05-04T08:30:00.000Z',
        },
      });
      expect(r.success, true);
      expect(r.flagEnabled, true);
      expect(r.data.goal, GoalCode.olympiad);
    });

    test('decodes flag-off envelope', () {
      final r = DailyPlanResponse.fromJson({
        'success': true,
        'flagEnabled': false,
        'data': {'goal': null, 'totalMinutes': 0, 'items': [], 'generatedAt': '2026-05-04T00:00:00.000Z'},
      });
      expect(r.flagEnabled, false);
      expect(r.data.isEmpty, true);
    });
  });
}

File: src/__tests__/goals/daily-plan.test.ts
/**
 * Tests for src/lib/goals/daily-plan.ts
 *
 * Verifies per-goal plan composition, bilingual P7 compliance, ±15% minutes
 * tolerance, deterministic clock injection, and the empty-shape contract for
 * null/unknown goals.
 */

import { describe, it, expect } from 'vitest';
import {
  buildDailyPlan,
  buildDailyPlanByCode,
  type DailyPlan,
  type DailyPlanItemKind,
} from '@/lib/goals/daily-plan';
import { GOAL_PROFILES, type GoalCode } from '@/lib/goals/goal-profile';

const ALL_GOALS: GoalCode[] = [
  'improve_basics',
  'pass_comfortably',
  'school_topper',
  'board_topper',
  'competitive_exam',
  'olympiad',
];

const DEVANAGARI_RE = /[ऀ-ॿ]/;
const ALLOWED_KINDS: ReadonlyArray<DailyPlanItemKind> = [
  'pyq',
  'concept',
  'practice',
  'challenge',
  'review',
  'reflection',
];

describe('buildDailyPlan / buildDailyPlanByCode', () => {
  describe('per-goal plan composition (matches authored spec)', () => {
    it.each(ALL_GOALS)('builds a non-empty plan for %s', (code) => {
      const profile = GOAL_PROFILES[code];
      const plan = buildDailyPlan(profile);
      expect(plan.goal).toBe(code);
      expect(plan.items.length).toBeGreaterThan(0);
      expect(plan.totalMinutes).toBe(
        plan.items.reduce((sum, it) => sum + it.estimatedMinutes, 0),
      );
    });

    it('improve_basics → 2 items totalling 10 min', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.improve_basics);
      expect(plan.items.length).toBe(2);
      expect(plan.totalMinutes).toBe(10);
      expect(plan.items.map((i) => i.kind)).toEqual(['concept', 'review']);
    });

    it('pass_comfortably → 3 items totalling 20 min', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.pass_comfortably);
      expect(plan.items.length).toBe(3);
      expect(plan.totalMinutes).toBe(20);
      expect(plan.items.map((i) => i.kind)).toEqual([
        'concept',
        'practice',
        'review',
      ]);
    });

    it('school_topper → 3 items totalling 30 min', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.school_topper);
      expect(plan.items.length).toBe(3);
      expect(plan.totalMinutes).toBe(30);
    });

    it('board_topper → 4 items totalling 45 min, includes pyq + reflection', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.board_topper);
      expect(plan.items.length).toBe(4);
      expect(plan.totalMinutes).toBe(45);
      expect(plan.items.map((i) => i.kind)).toEqual([
        'pyq',
        'practice',
        'review',
        'reflection',
      ]);
    });

    it('competitive_exam → 4 items totalling 60 min', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.competitive_exam);
      expect(plan.items.length).toBe(4);
      expect(plan.totalMinutes).toBe(60);
    });

    it('olympiad → 4 items totalling 60 min, includes challenge + reflection', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.olympiad);
      expect(plan.items.length).toBe(4);
      expect(plan.totalMinutes).toBe(60);
      expect(plan.items.map((i) => i.kind)).toContain('challenge');
      expect(plan.items.map((i) => i.kind)).toContain('reflection');
    });
  });

  describe('±15% minutes tolerance vs dailyTargetMinutes', () => {
    it.each(ALL_GOALS)('%s totalMinutes within ±15%% of dailyTargetMinutes', (code) => {
      const profile = GOAL_PROFILES[code];
      const plan = buildDailyPlan(profile);
      const target = profile.dailyTargetMinutes;
      const lower = target * 0.85;
      const upper = target * 1.15;
      expect(plan.totalMinutes).toBeGreaterThanOrEqual(lower);
      expect(plan.totalMinutes).toBeLessThanOrEqual(upper);
    });
  });

  describe('P7 bilingual + content rules', () => {
    it.each(ALL_GOALS)('%s: every item has non-empty en + hi labels', (code) => {
      const plan = buildDailyPlan(GOAL_PROFILES[code]);
      for (const item of plan.items) {
        expect(item.titleEn.length).toBeGreaterThan(0);
        expect(item.titleHi.length).toBeGreaterThan(0);
      }
    });

    it.each(ALL_GOALS)('%s: first item titleHi uses Devanagari script', (code) => {
      const plan = buildDailyPlan(GOAL_PROFILES[code]);
      expect(plan.items[0].titleHi).toMatch(DEVANAGARI_RE);
    });

    it.each(ALL_GOALS)('%s: every item kind is one of the allowed enum values', (code) => {
      const plan = buildDailyPlan(GOAL_PROFILES[code]);
      for (const item of plan.items) {
        expect(ALLOWED_KINDS).toContain(item.kind);
      }
    });

    it.each(ALL_GOALS)('%s: every rationale contains the goal code', (code) => {
      const plan = buildDailyPlan(GOAL_PROFILES[code]);
      for (const item of plan.items) {
        expect(item.rationale).toContain(`goal=${code}`);
        expect(item.rationale).toContain(`kind=${item.kind}`);
      }
    });
  });

  describe('buildDailyPlanByCode — null/unknown handling', () => {
    it('null returns empty shape', () => {
      const plan = buildDailyPlanByCode(null);
      expect(plan).toEqual<Partial<DailyPlan>>({
        goal: null,
        totalMinutes: 0,
        items: [],
        generatedAt: expect.any(String),
      });
    });

    it('undefined returns empty shape', () => {
      const plan = buildDailyPlanByCode(undefined);
      expect(plan.goal).toBeNull();
      expect(plan.items).toEqual([]);
      expect(plan.totalMinutes).toBe(0);
    });

    it('unknown goal code returns empty shape (no throw)', () => {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const plan = buildDailyPlanByCode('not_a_real_goal' as any);
      expect(plan.goal).toBeNull();
      expect(plan.items).toEqual([]);
    });

    it('valid goal code returns the same shape as buildDailyPlan(profile)', () => {
      const fixed = new Date('2026-05-03T12:00:00.000Z');
      const a = buildDailyPlanByCode('board_topper', { now: () => fixed });
      const b = buildDailyPlan(GOAL_PROFILES.board_topper, { now: () => fixed });
      expect(a).toEqual(b);
    });
  });

  describe('clock injection (test seam)', () => {
    it('honors opts.now for generatedAt', () => {
      const fixed = new Date('2026-01-15T08:30:00.000Z');
      const plan = buildDailyPlan(GOAL_PROFILES.board_topper, {
        now: () => fixed,
      });
      expect(plan.generatedAt).toBe('2026-01-15T08:30:00.000Z');
    });

    it('two builds with the same fixed clock are identical', () => {
      const fixed = new Date('2026-01-15T08:30:00.000Z');
      const a = buildDailyPlan(GOAL_PROFILES.olympiad, { now: () => fixed });
      const b = buildDailyPlan(GOAL_PROFILES.olympiad, { now: () => fixed });
      expect(a).toEqual(b);
    });

    it('without opts.now uses real time (ISO format)', () => {
      const plan = buildDailyPlan(GOAL_PROFILES.school_topper);
      expect(plan.generatedAt).toMatch(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/);
    });
  });
});
